### DISCUSSION TITLE
Promise integration

### DISCUSSION THREAD

**Discussion by kennyjwilli** [josdejong/mathjs#3234] [Q&A]
I am interested in integrating the [Promise POC](https://github.com/josdejong/mathjs/discussions/2665#discussioncomment-3414406) into my application using mathjs. Specifically, I would like to write an expression that concatenates a string together with a value returned from a promise-returning function. e.g., I would like to do something like this. 

```
concat('prefix--', generate_id())
  prefix--1
```

I have taken the POC code and written an example [JSFiddle here](https://jsfiddle.net/kennyjwilli/z2yafu54/). The short is, it doesn't seem to exactly work, and I'd love some help understand why, and if it is possible to fix. A basic example, similar to the POC example does work. 

```
concat('hello', getDelayedResult(' world')) 
     hello world
```

If I make the example slightly more complex by, say, having getDelayedResult return a number that I cast to a string, the promise gets embedded in the string. 

```
concat('hello', format(getDelayedResult(1))) 
     hello[object Promise]
```

The example code wraps the `format` function in the same promise resolving code as `concat`, but it does not seem to work for `format` for some reason. I thought it could be due to the nesting (e.g., concat -> format -> getDelayedResult), but removing the outer `concat` shows that it is unique to `format`. 

```
format(getDelayedResult(1)) 
     [object Promise]
```

And the same occurs with `string`.

```
string(getDelayedResult(1)) 
     [object Promise]
```

And just for completeness, I also tried with `format` with an object param. No dice. 

```
print('pre-$id', {id: getDelayedResult(1)}) 
     pre-[object Promise]
```

Any thoughts on what's going on here and how I may be able to fix/workaround this? 

--- 

Update: I found an interesting workaround -- [JSFiddle](https://jsfiddle.net/kennyjwilli/mecsz3du/6/). Essentially, if you update the type signature for the promise factory to `...Promise`, it will work for some of the examples. 

```
"concat('hello', format(getDelayedResult(1))) 
     hello[object Promise]"
"print('pre-$id', {id: getDelayedResult(1)}) 
     pre-[object Promise]"
"format(getDelayedResult(1)) 
     [object Promise]"
"concat('hello', getDelayedResult(' world')) 
     hello world"
"concat('hello', string(getDelayedResult(1))) 
     hello1"
"string(getDelayedResult(1)) 
     1"
```

**Comment by josdejong** [comment]
I remember there where some tricky edge cases with this async POC but I don't remember the details, see https://github.com/josdejong/mathjs/discussions/2665#discussioncomment-3414429

It would be interesting to pick up async support again BTW :)

**Reply by kennyjwilli** [reply]
Fwiw, my use case is similar to others that have brought up async -- I'd like to expose functionality from database interactions, which are exclusively async.
